passengerfile_parser.go

```diff
@@ -0,0 +1,47 @@
+package passenger
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+
+	"github.com/paketo-buildpacks/packit/v2/fs"
+)
+
+type PassengerfileParser struct{}
+
+func NewPassengerfileParser() PassengerfileParser {
+	return PassengerfileParser{}
+}
+
+type Passengerfile struct {
+	Port *int `json:"port,omitempty"`
+}
+
+func (p PassengerfileParser) ParsePort(path string, defaultPort int) (int, error) {
+	exists, err := fs.Exists(path)
+	if err != nil {
+		return 0, fmt.Errorf("failed to determine if Passengerfile exists: %w", err)
+	}
+
+	if !exists {
+		return defaultPort, nil
+	}
+
+	content, err := ioutil.ReadFile(path)
+	if err != nil {
+		return 0, fmt.Errorf("failed to read Passengerfile: %w", err)
+	}
+
+	passengerfile := Passengerfile{}
+	err = json.Unmarshal(content, &passengerfile)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse Passengerfile: %w", err)
+	}
+
+	if passengerfile.Port == nil {
+		return defaultPort, nil
+	}
+
+	return *passengerfile.Port, nil
```

Comment:
nit: can we squeeze this down a bit?

```suggestion
	file, err := os.Open(path)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return defaultPort, nil
		}

		return 0, fmt.Errorf("failed to read Passengerfile: %w", err)
	}
	defer file.Close()

	passengerfile := Passengerfile{Port: &defaultPort}
	err = json.NewDecoder(file).Decode(&passengerfile)
	if err != nil {
		return 0, fmt.Errorf("failed to parse Passengerfile: %w", err)
	}

	return *passengerfile.Port, nil
```

The prefilling of the `Port` value with the default might be a bit too clever, but it would let you change the type of the `Port` field to plain `int` instead of `*int`.

Reply:
Ok that's helpful, thanks.

>  I like APIs that hide the complexity of their implementation. If all I care about is knowing what port I should use for the server, then exposing the file content directly puts way more burden on the caller, and exposes unnecessary complexity.

Agreed, and I'm slightly confused because I don't know which of my points this is a reponse. But, I'm glad to hear we share the same high-level design goals.

> All of this being said, I'm fine with this code as-is. I'm just pointing out that I would have written this differently, and then trying to understand why that is.

Thanks - I appreciate the dialog. I've definitely learned things from it.